ndt7/server/upload/upload.go

```diff
@@ -0,0 +1,221 @@
+	t0 := time.Now()
+	ticker := time.NewTicker(spec.MinMeasurementInterval)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case now := <-ticker.C:
+			elapsed := now.Sub(t0)
+			if elapsed > defaultDuration {
+				logging.Logger.Debug("Upload has run for long enough.")
+				return
+			}
+
+			measurement := model.Measurement{
+				Elapsed: elapsed.Seconds(),
+			}
+
+			err = gatherAndSaveTCPInfoAndBBRInfo(&measurement, sockfp, resultfp)
+			if err != nil {
+				return // error printed already
+			}
+
+			dst <- measurement
+		}
+	}
+}
+
+// startMeasuring runs the measurement loop. This runs in a separate goroutine
+// and emits Measurement events on the returned channel.
+func startMeasuring(ctx context.Context, request *http.Request, conn *websocket.Conn, dataDir string) chan model.Measurement {
+	dst := make(chan model.Measurement)
+	go measuringLoop(ctx, request, conn, dataDir, dst)
+	return dst
+}
+
+// ServeHTTP handles the upload subtest.
+func (ul Handler) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	// TODO: factor out this stuff as it's the same for both UL/DL
+	logging.Logger.Debug("Upgrading to WebSockets")
+
+	if request.Header.Get("Sec-WebSocket-Protocol") != spec.SecWebSocketProtocol {
+		warnAndClose(writer, "Missing Sec-WebSocket-Protocol in request")
+		return
+	}
+	headers := http.Header{}
+	headers.Add("Sec-WebSocket-Protocol", spec.SecWebSocketProtocol)
+
+	conn, err := ul.Upgrader.Upgrade(writer, request, headers)
+	if err != nil {
+		warnAndClose(writer, "Cannot UPGRADE to WebSocket")
+		return
+	}
+
+	defer conn.Close()
+
+	// Read limit is set to the smallest allowed payload size.
+	conn.SetReadLimit(spec.MinMaxMessageSize)
+
+	// Start measuring loop
+	logging.Logger.Debug("Starting measuring goroutine")
+	ctx, cancel := context.WithCancel(request.Context())
+	measurements := startMeasuring(ctx, request, conn, ul.DataDir)
+
+	// Make sure we clean up resources
+	defer func() {
+		logging.Logger.Debug("Closing the WebSocket connection")
+		conn.WriteControl(websocket.CloseMessage, websocket.FormatCloseMessage(
+			websocket.CloseNormalClosure, ""), time.Now().Add(defaultTimeout))
+
+		// We could leave the context because the measuring goroutine thinks we're
+		// done or because there has been a socket error. In the latter case, it is
+		// important to synchronise with the goroutine and wait for it to exit.
+		cancel()
+		for range measurements {
+			// NOTHING
+		}
+	}()
+
+	logging.Logger.Debug("Starting receiving data from the client")
+	for {
+		select {
+		case _, ok := <-measurements:
+			if !ok {
+				return // the goroutine told us it's time to stop running
+			}
+		default:
+			conn.SetReadDeadline(time.Now().Add(defaultTimeout))
+			mt, message, err := conn.ReadMessage()
+			if err != nil {
+				if !websocket.IsCloseError(err, websocket.CloseNormalClosure) {
+					logging.Logger.WithError(err)
+				}
+				logging.Logger.Debug(request.RemoteAddr + ": connection closed.")
```

Comment:
I believe this log message is redundant as we already log in the `defer func()` above